File: services/menos/menos/auth/signature.py
"""RFC 9421 HTTP Message Signature verification."""

import hashlib
import re
from base64 import b64decode
from datetime import UTC, datetime

from cryptography.exceptions import InvalidSignature
from fastapi import HTTPException, Request

from menos.auth.keys import KeyStore


class SignatureVerifier:
    """Verifies HTTP Message Signatures per RFC 9421."""

    # Signature validity window (seconds)
    MAX_AGE = 300  # 5 minutes

    def __init__(self, key_store: KeyStore):
        self.key_store = key_store

    async def verify_request(self, request: Request) -> str:
        """Verify request signature, return key_id if valid."""
        sig_input = request.headers.get("signature-input")
        signature = request.headers.get("signature")

        if not sig_input or not signature:
            raise HTTPException(401, "Missing signature headers")

        params = self._parse_signature_input(sig_input)
        key_id = self._validate_params(params)
        self._check_timestamp(params.get("created"))

        public_key = self.key_store.get_key(key_id)
        if not public_key:
            raise HTTPException(401, f"Unknown key: {key_id}")

        covered_components = params.get("components", [])
        signature_base = await self._build_signature_base(request, covered_components, sig_input)
        sig_bytes = self._extract_signature(signature)

        try:
            public_key.verify(sig_bytes, signature_base.encode())
        except InvalidSignature:
            raise HTTPException(401, "Invalid signature")

        return key_id

    def _validate_params(self, params: dict) -> str:
        """Validate signature parameters and return key_id."""
        key_id = params.get("keyid")
        if not key_id:
            raise HTTPException(401, "Missing keyid in signature-input")
        alg = params.get("alg", "ed25519")
        if alg != "ed25519":
            raise HTTPException(401, f"Unsupported algorithm: {alg}")
        return key_id

    def _check_timestamp(self, created: str | None) -> None:
        """Raise 401 if the created timestamp is outside the validity window."""
        if not created:
            return
        created_time = datetime.fromtimestamp(int(created), tz=UTC)
        age = (datetime.now(UTC) - created_time).total_seconds()
        if abs(age) > self.MAX_AGE:
            raise HTTPException(401, "Signature expired or from future")

    def _parse_signature_input(self, sig_input: str) -> dict:
        """Parse signature-input header."""
        # Format: sig1=("@method" "@path" ...);keyid="...";created=...;alg="ed25519"
        result = {"components": []}

        # Extract label and value
        match = re.match(r"^(\w+)=\(([^)]*)\);?(.*)$", sig_input)
        if not match:
            raise HTTPException(400, "Invalid signature-input format")

        label, components_str, params_str = match.groups()

        # Parse components
        for comp in re.findall(r'"([^"]+)"', components_str):
            result["components"].append(comp)

        # Parse parameters
        for param_match in re.finditer(r'(\w+)=(?:"([^"]+)"|(\d+))', params_str):
            key = param_match.group(1)
            value = param_match.group(2) or param_match.group(3)
            result[key] = value

        return result

    async def _build_signature_base(
        self, request: Request, components: list[str], sig_input: str
    ) -> str:
        """Build the signature base string per RFC 9421."""
        lines = [await self._resolve_component(request, c) for c in components]
        sig_params = sig_input.split("=", 1)[1] if "=" in sig_input else sig_input
        lines.append(f'"@signature-params": {sig_params}')
        return "\n".join(lines)

    async def _resolve_component(self, request: Request, component: str) -> str:
        """Resolve a single signature component to its header line."""
        if component == "@method":
            return f'"@method": {request.method}'
        if component == "@path":
            path = request.url.path
            if request.url.query:
                path = f"{path}?{request.url.query}"
            return f'"@path": {path}'
        if component == "@authority":
            return f'"@authority": {request.headers.get("host", "")}'
        if component == "@target-uri":
            return f'"@target-uri": {str(request.url)}'
        if component == "content-digest":
            body = await request.body()
            digest = hashlib.sha256(body).digest()
            digest_b64 = __import__("base64").b64encode(digest).decode()
            return f'"content-digest": sha-256=:{digest_b64}:'
        # Regular header
        value = request.headers.get(component, "")
        return f'"{component}": {value}'

    def _extract_signature(self, signature_header: str) -> bytes:
        """Extract signature bytes from header."""
        # Format: sig1=:base64data:
        match = re.match(r"^\w+=:([A-Za-z0-9+/=]+):", signature_header)
        if not match:
            raise HTTPException(400, "Invalid signature format")
        return b64decode(match.group(1))


async def verify_signature(request: Request, key_store: KeyStore) -> str:
    """Verify request signature, return key_id."""
    verifier = SignatureVerifier(key_store)
    return await verifier.verify_request(request)

File: services/core/src/vault/signature.ts
import { createHash, createPublicKey, verify as cryptoVerify } from "node:crypto";
import { HttpError } from "./errors";
import type { KeyStore } from "./keys";

/** DER prefix that wraps a raw ed25519 public key into SPKI format. */
const ED25519_SPKI_PREFIX = Buffer.from("302a300506032b6570032100", "hex");

/** Signature validity window in seconds. */
const MAX_AGE_SECONDS = 300;

/**
 * Request shape consumed by signature verification, decoupled from the HTTP
 * server implementation. `path` must include the query string when present,
 * `headers` must use lowercase names, and `body` must be the raw bytes.
 */
export type SignedRequestInput = {
  method: string;
  path: string;
  targetUri?: string;
  headers: Record<string, string | undefined>;
  body?: Buffer;
};

type SignatureParams = {
  components: string[];
  params: Record<string, string>;
  signatureParams: string;
};

const REQUIRED_COMPONENTS = ["@method", "@path", "@authority"] as const;

function parseSignatureInput(sigInput: string): SignatureParams {
  const match = /^(\w+)=\(([^)]*)\);?(.*)$/.exec(sigInput);
  if (match === null) {
    throw new HttpError(400, "Invalid signature-input format");
  }
  const componentsStr = match[2] ?? "";
  const paramsStr = match[3] ?? "";
  const components: string[] = [];
  for (const component of componentsStr.matchAll(/"([^"]+)"/g)) {
    if (component[1] !== undefined) components.push(component[1]);
  }
  const params: Record<string, string> = {};
  for (const param of paramsStr.matchAll(/(\w+)=(?:"([^"]+)"|(\d+))/g)) {
    const key = param[1];
    const value = param[2] ?? param[3];
    if (key !== undefined && value !== undefined) params[key] = value;
  }
  const separatorIndex = sigInput.indexOf("=");
  const signatureParams = separatorIndex >= 0 ? sigInput.slice(separatorIndex + 1) : sigInput;
  return { components, params, signatureParams };
}

function validateParams(params: Record<string, string>): string {
  const keyId = params.keyid;
  if (keyId === undefined || keyId === "") {
    throw new HttpError(401, "Missing keyid in signature-input");
  }
  const alg = params.alg ?? "ed25519";
  if (alg !== "ed25519") {
    throw new HttpError(401, `Unsupported algorithm: ${alg}`);
  }
  return keyId;
}

function checkTimestamp(created: string | undefined): void {
  if (created === undefined) {
    throw new HttpError(401, "Missing created in signature-input");
  }
  if (!/^\d+$/.test(created)) {
    throw new HttpError(401, "Invalid created in signature-input");
  }
  const createdSeconds = Number(created);
  const ageSeconds = Date.now() / 1000 - createdSeconds;
  if (!Number.isSafeInteger(createdSeconds) || !Number.isFinite(ageSeconds) || Math.abs(ageSeconds) > MAX_AGE_SECONDS) {
    throw new HttpError(401, "Signature expired or from future");
  }
}

function contentDigest(body: Buffer | undefined): string {
  const digest = createHash("sha256").update(body ?? Buffer.alloc(0)).digest("base64");
  return `sha-256=:${digest}:`;
}

function validateComponentProfile(request: SignedRequestInput, components: string[]): void {
  const bodyIsNonempty = (request.body?.length ?? 0) > 0;
  const requiredComponents = bodyIsNonempty
    ? [...REQUIRED_COMPONENTS, "content-digest"]
    : REQUIRED_COMPONENTS;
  if (
    components.length !== requiredComponents.length
    || components.some((component, index) => component !== requiredComponents[index])
  ) {
    throw new HttpError(401, "Invalid signature component profile");
  }

  const suppliedDigest = request.headers["content-digest"];
  if (bodyIsNonempty && suppliedDigest === undefined) {
    throw new HttpError(401, "Missing content-digest header");
  }
  if (suppliedDigest !== undefined && suppliedDigest !== contentDigest(request.body)) {
    throw new HttpError(401, "Invalid content-digest header");
  }
}

function resolveComponent(request: SignedRequestInput, component: string): string {
  if (component === "@method") {
    return `"@method": ${request.method}`;
  }
  if (component === "@path") {
    return `"@path": ${request.path}`;
  }
  if (component === "@authority") {
    return `"@authority": ${request.headers.host ?? ""}`;
  }
  if (component === "@target-uri") {
    return `"@target-uri": ${request.targetUri ?? ""}`;
  }
  if (component === "content-digest") {
    return `"content-digest": ${request.headers["content-digest"] ?? ""}`;
  }
  return `"${component}": ${request.headers[component] ?? ""}`;
}

function buildSignatureBase(
  request: SignedRequestInput,
  components: string[],
  signatureParams: string,
): string {
  const lines = components.map((component) => resolveComponent(request, component));
  lines.push(`"@signature-params": ${signatureParams}`);
  return lines.join("\n");
}

function extractSignature(signatureHeader: string): Buffer {
  const match = /^\w+=:([A-Za-z0-9+/=]+):/.exec(signatureHeader);
  if (match === null || match[1] === undefined) {
    throw new HttpError(400, "Invalid signature format");
  }
  return Buffer.from(match[1], "base64");
}

/**
 * Verifies an RFC 9421 HTTP message signature and returns the authenticated
 * key id. Semantics mirror the Python Menos verifier so signatures produced
 * by the unmodified dotfiles client signer keep verifying.
 */
export function verifySignedRequest(request: SignedRequestInput, keyStore: KeyStore): string {
  const sigInput = request.headers["signature-input"];
  const signature = request.headers.signature;
  if (sigInput === undefined || sigInput === "" || signature === undefined || signature === "") {
    throw new HttpError(401, "Missing signature headers");
  }

  const parsed = parseSignatureInput(sigInput);
  const keyId = validateParams(parsed.params);
  checkTimestamp(parsed.params.created);
  validateComponentProfile(request, parsed.components);

  const publicKeyRaw = keyStore.getKey(keyId);
  if (publicKeyRaw === undefined) {
    throw new HttpError(401, `Unknown key: ${keyId}`);
  }

  const signatureBase = buildSignatureBase(request, parsed.components, parsed.signatureParams);
  const signatureBytes = extractSignature(signature);
  const publicKey = createPublicKey({
    key: Buffer.concat([ED25519_SPKI_PREFIX, publicKeyRaw]),
    format: "der",
    type: "spki",
  });

  const valid = cryptoVerify(null, Buffer.from(signatureBase, "utf8"), publicKey, signatureBytes);
  if (!valid) {
    throw new HttpError(401, "Invalid signature");
  }
  return keyId;
}
